File: partiql-ast/src/main/java/org/partiql/ast/v1/Sort.java
package org.partiql.ast.v1;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.ast.v1.expr.Expr;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;

/**
 * TODO docs, equals, hashcode
 */
public class Sort extends AstNode {
    @NotNull
    public Expr expr;

    @Nullable
    public Order order;

    @Nullable
    public Nulls nulls;

    public Sort(@NotNull Expr expr, @Nullable Order order, @Nullable Nulls nulls) {
        this.expr = expr;
        this.order = order;
        this.nulls = nulls;
    }

    @NotNull
    @Override
    public Collection<AstNode> children() {
        List<AstNode> kids = new ArrayList<>();
        kids.add(expr);
        return kids;
    }

    @Override
    public <R, C> R accept(@NotNull AstVisitor<R, C> visitor, C ctx) {
        return visitor.visitSort(this, ctx);
    }
}

File: partiql-ast/src/main/kotlin/org/partiql/ast/v1/Sort.kt
package org.partiql.ast.v1

/**
 * TODO docs, equals, hashcode
 */
public class Sort(
    @JvmField
    public var expr: Expr,
    @JvmField
    public var dir: Dir?,
    @JvmField
    public var nulls: Nulls?,
) : AstNode() {
    public override fun children(): Collection<AstNode> {
        val kids = mutableListOf<AstNode?>()
        kids.add(expr)
        return kids.filterNotNull()
    }

    public override fun <R, C> accept(visitor: AstVisitor<R, C>, ctx: C): R = visitor.visitSort(
        this,
        ctx
    )

    /**
     * TODO docs, equals, hashcode
     */
    public enum class Dir {
        ASC,
        DESC,
        OTHER,
    }

    /**
     * TODO docs, equals, hashcode
     */
    public enum class Nulls {
        FIRST,
        LAST,
        OTHER,
    }
}
